addons/pos_online_payment/controllers/payment_portal.py

```diff
@@ -0,0 +1,302 @@
+            raise ValidationError(
+                _("The amount to pay has changed. Please refresh the page."))
+
+        payment_option_id = kwargs.get('payment_option_id', None)
+        if not payment_option_id:
+            raise UserError(_("A payment option must be specified."))
+        flow = kwargs.get('flow', None)
+        if not flow or flow not in ['redirect', 'direct', 'token']:
+            raise UserError(_("The payment should either be direct, with redirection, or made by a token."))
+        providers_sudo = self._get_allowed_providers_sudo(request, pos_order_sudo, partner_sudo.id, amount_to_pay)
+        if flow == 'token':
+            tokens_sudo = request.env['payment.token']._get_available_tokens(
+                providers_sudo.ids, partner_sudo.id)
+            if payment_option_id not in tokens_sudo.ids:
+                raise UserError(_("The payment token is invalid."))
+        else:
+            if payment_option_id not in providers_sudo.ids:
+                raise UserError(_("The payment provider is invalid."))
+
+        kwargs['reference_prefix'] = None  # Computed with pos_order_id
+        kwargs.pop('pos_order_id', None) # _create_transaction kwargs keys must be different than custom_create_values keys
+
+        tx_sudo = self._create_transaction(**kwargs)
+        tx_sudo.landing_route = self._get_landing_route(pos_order_sudo.id, access_token, exit_route_arg=self._get_exit_route_arg(exit_route), tx_id=tx_sudo.id)
+
+        return tx_sudo._get_processing_values()
+
+    @http.route('/pos/pay/confirmation/<int:pos_order_id>', type='http', methods=['GET'], auth='public', website=True, sitemap=False)
+    def pos_order_pay_confirmation(self, pos_order_id=None, tx_id=None, access_token=None, exit_route=None, **kwargs):
+        """ Behaves like payment.PaymentPortal.payment_confirm but for POS online payment.
+
+        :param str pos_order_id: The POS order to confirm, as a `pos.order` id
+        :param str tx_id: The transaction to confirm, as a `payment.transaction` id
+        :param str access_token: The access token used to verify the user
+        :param str exit_route: The URL to open to leave the POS online payment flow
+        :param dict kwargs: Data from payment module
+
+        :return: The rendered confirmation page
+        :rtype: str
+        :raise: AccessError if the provided order or access token is invalid
+        """
+        tx_id = self._cast_as_int(tx_id)
+        rendering_context = {
+            'state': 'error',
+            'exit_route': exit_route,
+            'pay_route': self._get_pay_route(pos_order_id, access_token, exit_route)
+        }
+        if not tx_id or not pos_order_id:
+            return self._render_pay_confirmation(request, rendering_context)
+
+        pos_order_sudo = self._check_order_access(pos_order_id, access_token)
+
+        tx_sudo = request.env['payment.transaction'].sudo().browse(tx_id)
```

Comment:
```suggestion
        tx_sudo = request.env['payment.transaction'].sudo().search([('id', '=', tx_id)])
```
When you use `browse`, you probably need to call `.exists()` on it. E.g., you can do "browse(99999)" and it will just immediately create the object, without checking the db. (Something I observed recently.)

That's probably the reason why you need to handle `MissingError` below.

Reply:
Ok thank you, changed it